src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -1634,46 +1639,105 @@ private bool DoesParamArgRankMatch(List<Type> parameterTypes, List<AssociativeNo
+                {
+                    Type t;
+                    if (!astTypeInfoMap.TryGetValue(rgnNode.ID, out t))
                     {
-                        var argRank = GetRank(t);
-                        var paramRank = GetRank(p);
-                        if (argRank != paramRank) return false;
-
-                        // If both have rank 0, it could also be because their type info is ambiguous.
-                        if (argRank == 0 && paramRank == 0) return false;
+                        throw new Exception("unkown ast type");
                     }
+                    if (!DoesParamRankMatchArg(fep,p, t,argIndex)) return false;
                 }
                 else return false;
                 argIndex++;
             }
             return true;
         }
 
-        private static int GetRank(Type type)
+        /// <summary>
+        /// Check that a parameter's rank and type are compatible with the given argument.
+        /// This function besides checking rank directly, also does checks using the FEP attributes.
+        /// </summary>
+        /// <param name="fep">the function end point wrapper</param>
+        /// <param name="paramType">type of param on the fep</param>
+        /// <param name="argType">actual argument type</param>
+        /// <param name="paramIndex">parameter index in list of params to fep.</param>
+        /// <returns></returns>
+        private static bool DoesParamRankMatchArg(ProtoCore.CLRFunctionEndPoint fep,Type paramType, Type argType, int paramIndex)
+        {
+            if (!paramType.Equals(argType))
+            {
+                //if both types are value types, let's
+                //let coercion figure this out.
+                if (paramType.IsValueType && argType.IsValueType)
+                {
+                    return true;
+                }
+
+                var argRank = GetRank(fep,argType,paramIndex);
+                var paramRank = GetRank(fep,paramType, paramIndex);
+
+                //if the param is an arbitrary rank array and
+                //arg is some array type, replication should not occur.
+                if (paramRank == -1 && argRank > 0)
+                {
+                    return true;
+                }
+                if (argRank != paramRank) return false;
+
+                // If both have rank 0, it could also be because their type info is ambiguous.
+                //TODO commenting this out will makes negate test pass - as negate method expects object param.
+                //TODO perhaps just log something
+                //if (argRank == 0 && paramRank == 0) return false;
+
+                if (!paramType.IsAssignableFrom(argType))
+                {
+                    return false;
+                }
+
+            }
+            return true;
+        }
+
+        private static int GetRank(ProtoCore.CLRFunctionEndPoint fep,Type type,int paramIndex)
         {
+            //TODO (Alternatives to consider)
```

Comment:
I'd like some feedback on these alternatives
1. we use the imported protocore typesystem and just check if the imported rank is arbitrary
2. we try to minimally recreate the logic that the type system encodes to try to remove our dependency on it.

Reply:
Well we did have discussions about removing the dependency on the old VM.
Not sure what that means now, maybe we can share some of that logic or re-write it.
I think we can use the protocore system now and we can separate out later